src/notebooks/controllers/installPythonControllerCommands.ts

```diff
@@ -161,4 +176,40 @@ export class InstallPythonControllerCommands implements IExtensionSingleActivati
             }
         }
     }
+
+    // We don't always want to show our modal warning for installing the python extension
+    // this function will choose if this should be shown, and return true if the install should
+    // proceed and false otherwise
+    private async shouldInstallExtensionPrompt(): Promise<boolean> {
+        // We want to show the dialog if the active document is running, in this case, the command
+        // was triggered from the run button and we want to warn the user what we are doing
+        if (this.isActiveNotebookDocumentRunning()) {
+            // First present a simple modal dialog to indicate what we are about to do
+            const selection = await this.appShell.showInformationMessage(
+                DataScience.pythonExtensionRequiredToRunNotebook(),
+                { modal: true },
+                Common.install()
+            );
+            if (selection === Common.install()) {
+                sendTelemetryEvent(Telemetry.PythonExtensionNotInstalled, undefined, { action: 'download' });
+                return true;
+            } else {
+                // If they don't want to install, just bail out at this point
+                sendTelemetryEvent(Telemetry.PythonExtensionNotInstalled, undefined, { action: 'dismissed' });
+                return false;
+            }
+        }
+
+        // If the active notebook is not running, this command was triggered by direction clicking
+        // in this case, they clicked on "Install Python Extension" so no need for a modal
+        return true;
+    }
+    private isActiveNotebookDocumentRunning(): boolean {
+        if (window.activeNotebookEditor) {
+            return window.activeNotebookEditor.notebook.getCells().some((cell) => {
```

Comment:
Can we not track the notebook instead of tracking the cells?
i.e. `WeakSet<NotebookDocument>`. Then we can get rid of the iterate through all the cells to check if the notebook is running

Reply:
Nope won't work